helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java

```diff
@@ -429,24 +429,31 @@ private void updateInstanceSets(Map<String, InstanceConfig> instanceConfigMap,
     }
 
     liveInstancesMap.forEach((instanceName, liveInstance) -> {
-      if (_assignableInstanceConfigMap.containsKey(instanceName)) {
-        _assignableLiveInstancesMap.put(instanceName, liveInstance);
+      if (newAssignableInstanceConfigMap.containsKey(instanceName)) {
+        newAssignableLiveInstancesMap.put(instanceName, liveInstance);
       }
     });
 
     swapOutLogicalIdsByInstanceName.forEach((swapOutInstanceName, value) -> {
       String swapInInstanceName = swapInInstancesByLogicalId.get(value);
       if (swapInInstanceName != null) {
-        _swapOutInstanceNameToSwapInInstanceName.put(swapOutInstanceName, swapInInstanceName);
+        newSwapOutInstanceNameToSwapInInstanceName.put(swapOutInstanceName, swapInInstanceName);
         if (liveInstancesMap.containsKey(swapInInstanceName)) {
-          _liveSwapInInstanceNames.add(swapInInstanceName);
+          newLiveSwapInInstanceNames.add(swapInInstanceName);
         }
         if (InstanceValidationUtil.isInstanceEnabled(instanceConfigMap.get(swapInInstanceName),
             clusterConfig)) {
-          _enabledSwapInInstanceNames.add(swapInInstanceName);
+          newEnabledSwapInInstanceNames.add(swapInInstanceName);
         }
       }
     });
+
+    // Replace caches with up to date instance sets.
```

Comment:
question: do we want to have some kind of barrier so that when user access different maps, they will access the same snapshot. (i.e., not the new _assignableInstanceConfigMap and old _enabledSwapInInstanceNames etc)

Reply:
I have created new class called DerivedInstanceCache. We will swap out this object which encapsulates the maps and sets so it will update as a snapshot.